# Project 1: Traveling Salesperson Problem (TSP)<br><small>Joshua Nasiatka (2016.11.13)

## Problem 1: 
This problem solves a real-world Traveling Salesperson Problem (TSP) using distance data. We will find the shortest total path between the Massachusetts cities/towns: Boston, Fall River, Fitchburg, Lowell, North Adams, Springfield and Worcester. Follow these steps to do this:  

**( a ) Use the brute force algorithm that goes through each path to determine the shortest total distance. Find the distance and write down the path as it goes through each town. Start and end with Fitchburg and write down the town names of the shortest path.**

In [1]:
push!(LOAD_PATH,".");

In [2]:
using TSP

I cannot fathom how to go about the round-trip Fitchburg => (cities) => Fitchburg with checking permutations, I'm probably missing something simple, I just can't figure out what it is. This goes for most of the questions; I'm not sure how to explicitly check for Fitchburg as Starting position/location. So instead, just manually wrote out the path from Fitchburg based on the outputted shortest path.

In [3]:
min_dist, tpath, dist_map = run_TSP("map_q1.json");

The Shortest Distance and Position is: (596457.0,2930)

The path of which is: 
String["North Adams, MA 01247, USA","Fitchburg, MA 01420, USA","Lowell, MA, USA","Boston, MA, USA","Fall River, MA, USA","Worcester, MA, USA","Springfield, MA, USA","North Adams, MA 01247, USA"]

As for a round trip from Fitchburg:  
Fitchburg => Lowell => Boston => Fall River => Worcester => Springfield => North Adams => Fitchburg

In [4]:
0.000621371*dist_map

7×7 Array{Float64,2}:
  0.0      47.0539   83.0232   31.4153   87.2026   78.7066   27.2366
 52.9017    0.0      51.8894   30.3484  136.016    91.1992   47.1925
 84.1908   52.3747    0.0      77.0351  169.439   102.106    56.9598
 33.1967   30.121    77.0469    0.0     116.311    93.7052   42.2358
 86.895   130.658   157.372   115.02      0.0      61.1441  107.684 
 79.6504   90.9302  102.107    92.8732   73.7605    0.0      52.4195
 27.83     47.5529   56.5124   41.0527  119.529    52.1964    0.0   

In [5]:
tpath

8-element Array{String,1}:
 "North Adams, MA 01247, USA"
 "Fitchburg, MA 01420, USA"  
 "Lowell, MA, USA"           
 "Boston, MA, USA"           
 "Fall River, MA, USA"       
 "Worcester, MA, USA"        
 "Springfield, MA, USA"      
 "North Adams, MA 01247, USA"

**( b ) Add the following towns: Amherst, Braintree, Lawrence, Newton, Natick, Pittsfield, Plymouth, Provincetown, Salem, Woburn to the list above. Produce an array of distances in alphabetical order.**

In [6]:
A = render_map("map_q1.json");
B = render_map("map_q2.json");
C = render_map("map_q3.json");
D = render_map("map_q4.json");

In [7]:
the_map = 0.000621371*[A C; B D]

17×17 Array{Float64,2}:
   0.0      47.0539   83.0232   31.4153  …  162.206    59.7454   44.3963
  52.9017    0.0      51.8894   30.3484     115.131    22.1519   11.857 
  84.1908   52.3747    0.0      77.0351     103.698    73.246    62.9517
  33.1967   30.121    77.0469    0.0        147.455    32.996    17.6469
  86.895   130.658   157.372   115.02       245.81    143.349   128.0   
  79.6504   90.9302  102.107    92.8732  …  195.148   108.905    93.5561
  27.83     47.5529   56.5124   41.0527     155.498    65.5279   50.1782
  49.5929   93.6257  104.802    77.7174     197.843   106.047    90.6978
  61.0733   13.3582   42.8976   44.9941     103.208    34.2301   23.9352
  44.6225   29.6295   80.8273   16.7907     144.068    32.8457   22.3047
  43.8632   10.9796   52.9557   27.7846  …  123.364    34.9931   19.644 
  44.7151   24.1247   54.992    34.5731     125.4      41.7816   26.4319
  99.6381  138.095   149.271   140.037      242.312   156.069   140.72  
  88.2838   40.5339   37.58

In [8]:
#run_TSP(the_map)

This doesn't work due to lack of resources, requires way (way) more than the average computer.  
17! permutations...uhh...no.

**( c ) Use a Simulated Annealing algorithm to find a minimum path through all of the towns in `(a)` and `(b)`. Use as large a number of randomizations in your algorithm as possible. Find the total distance as well as the path of the towns. Start and end with Fitchburg and write down the town names of the shortest path.**

In [9]:
rpath=nthperm(collect(1:17),rand(1:factorial(8)))

17-element Array{Int64,1}:
  1
  2
  3
  4
  5
  6
  7
  8
  9
 12
 14
 13
 17
 16
 10
 11
 15

In [10]:
current_dist=total_distance(rpath, the_map)

1320.7880617130004

In [11]:
swapEls!(rpath,1,4)

17-element Array{Int64,1}:
  4
  2
  3
  1
  5
  6
  7
  8
  9
 12
 14
 13
 17
 16
 10
 11
 15

In [12]:
new_path = swapEls!(rpath,rand(1:17),rand(1:17))

17-element Array{Int64,1}:
  4
  2
  3
  1
  5
  7
  6
  8
  9
 12
 14
 13
 17
 16
 10
 11
 15

In [13]:
total_distance(new_path,the_map)

1285.2530969650002

In [14]:
sim_annealing(the_map,10^9)

([13,1,8,3,16,2,17,11,12,4,10,5,7,15,6,9,14],707.9286016710001)

In [15]:
@time out = sim_annealing(the_map,10^8)

  7.732621 seconds (135 allocations: 8.063 KB)


([14,8,11,6,13,12,17,9,10,5,1,2,16,3,15,4,7],732.8126461079999)

```
1  => Fitchburg, MA
2  => Boston, MA
3  => Fall River, MA
4  => Lowell, MA
5  => North Adams, MA
6  => Springfield, MA
7  => Worcester, MA
8  => Amherst, MA
9  => Braintree, MA
10 => Lawrence, MA
11 => Newton, MA
12 => Natick, MA
13 => Pittsfield, MA
14 => Plymouth, MA
15 => Provincetown, MA
16 => Salem, MA
17 => Woburn, MA
```

**Having got the following as an output:  **
`([14,6,4,1,5,2,8,10,7,11,13,17,12,15,16,3,9],726.598936108)`  

**The shortest path by simulated annealing is:**  
`Plymouth => Springfield => Lowell => Fitchburg => North Adams => Boston => Amherst => Lawrence => Worcester => Newton => Pittsfield => Woburn => Natick => Provincetown => Salem => Fall River => Braintree`

<img src="map_output.png" width="800px">

That path is ridiculous and doesn't make sense as instead of a circle it is weaving across the state. Something either went direly wrong or is missing.

Hmm......

## Problem 2:
Consider the following weighted graph:
<img src="graph01.png" width=250 />
Use the method from class using tropical arithmetic on matrices to find the shortest path from A to H.

In [16]:
w = [0   9   Inf 13  16  Inf Inf Inf;
     9   0   15  Inf 3   Inf Inf Inf;
     Inf 15  0   Inf 10  4   Inf Inf;
     13  Inf Inf 0   18  Inf Inf Inf;
     16  3   10  18  0   17  21  Inf;
     Inf Inf 4   Inf 17  0   12  24;
     Inf Inf Inf Inf 21  12  0   5;
     Inf Inf Inf Inf Inf 24  5   0]

8×8 Array{Float64,2}:
   0.0    9.0  Inf     13.0   16.0  Inf    Inf    Inf  
   9.0    0.0   15.0  Inf      3.0  Inf    Inf    Inf  
 Inf     15.0    0.0  Inf     10.0    4.0  Inf    Inf  
  13.0  Inf    Inf      0.0   18.0  Inf    Inf    Inf  
  16.0    3.0   10.0   18.0    0.0   17.0   21.0  Inf  
 Inf    Inf      4.0  Inf     17.0    0.0   12.0   24.0
 Inf    Inf    Inf    Inf     21.0   12.0    0.0    5.0
 Inf    Inf    Inf    Inf    Inf     24.0    5.0    0.0

In [17]:
using TropicalMath

In [18]:
w⊗w

8×8 Array{Float64,2}:
   0.0    9.0  24.0   13.0  12.0  33.0  37.0  Inf  
   9.0    0.0  13.0   21.0   3.0  19.0  24.0  Inf  
  24.0   13.0   0.0   28.0  10.0   4.0  16.0   28.0
  13.0   21.0  28.0    0.0  18.0  35.0  39.0  Inf  
  12.0    3.0  10.0   18.0   0.0  14.0  21.0   26.0
  33.0   19.0   4.0   35.0  14.0   0.0  12.0   17.0
  37.0   24.0  16.0   39.0  21.0  12.0   0.0    5.0
 Inf    Inf    28.0  Inf    26.0  17.0   5.0    0.0

In [19]:
1.0*((w⊗w)⊗w)⊗w

8×8 Array{Float64,2}:
  0.0   9.0  22.0  13.0  12.0  26.0  33.0  38.0
  9.0   0.0  13.0  21.0   3.0  17.0  24.0  29.0
 22.0  13.0   0.0  28.0  10.0   4.0  16.0  21.0
 13.0  21.0  28.0   0.0  18.0  32.0  39.0  44.0
 12.0   3.0  10.0  18.0   0.0  14.0  21.0  26.0
 26.0  17.0   4.0  32.0  14.0   0.0  12.0  17.0
 33.0  24.0  16.0  39.0  21.0  12.0   0.0   5.0
 38.0  29.0  21.0  44.0  26.0  17.0   5.0   0.0

^ These are the shortest distances with a max step size of 3  
The shortest distance from A => H is 38

A => B => E => G => H  
9 + 3 + 21 + 5 = 38

## Problem 3:
Use the Dijkstra’s Algorithm to find the shortest path from A to H for the graph in the previous problem.

In [20]:
using Graphs;

In [21]:
g = simple_inclist(8,is_directed=false)

Undirected Graph (8 vertices, 0 edges)

In [22]:
A = 1
B = 2
C = 3
D = 4
E = 5
F = 6
G = 7
H = 8
inputs = [
    (A, B, 9.0),
    (A, E, 16.0),
    (A, D, 13.0),
    (B, A, 9.0),
    (B, E, 3.0),
    (B, C, 15.0),
    (C, B, 15.0),
    (C, E, 10.0),
    (C, F, 4.0),
    (D, A, 13.0),
    (D, E, 18.0),
    (E, D, 18.0),
    (E, A, 16.0),
    (E, B, 3.0),
    (E, C, 10.0),
    (E, F, 17.0),
    (E, G, 21.0),
    (F, C, 4.0),
    (F, E, 17.0),
    (F, G, 12.0),
    (F, H, 24.0),
    (G, E, 21.0),
    (G, F, 12.0),
    (G, H, 5.0),
    (H, F, 24.0),
    (H, G, 5.0)
]

26-element Array{Tuple{Int64,Int64,Float64},1}:
 (1,2,9.0) 
 (1,5,16.0)
 (1,4,13.0)
 (2,1,9.0) 
 (2,5,3.0) 
 (2,3,15.0)
 (3,2,15.0)
 (3,5,10.0)
 (3,6,4.0) 
 (4,1,13.0)
 (4,5,18.0)
 (5,4,18.0)
 (5,1,16.0)
 (5,2,3.0) 
 (5,3,10.0)
 (5,6,17.0)
 (5,7,21.0)
 (6,3,4.0) 
 (6,5,17.0)
 (6,7,12.0)
 (6,8,24.0)
 (7,5,21.0)
 (7,6,12.0)
 (7,8,5.0) 
 (8,6,24.0)
 (8,7,5.0) 

In [23]:
dist = map(a->a[3],inputs)

26-element Array{Float64,1}:
  9.0
 16.0
 13.0
  9.0
  3.0
 15.0
 15.0
 10.0
  4.0
 13.0
 18.0
 18.0
 16.0
  3.0
 10.0
 17.0
 21.0
  4.0
 17.0
 12.0
 24.0
 21.0
 12.0
  5.0
 24.0
  5.0

In [24]:
for edge in inputs
    add_edge!(g,edge[1],edge[2])
end

In [25]:
adjacency_matrix(g)

8×8 Array{Bool,2}:
 false   true  false   true   true  false  false  false
  true  false   true  false   true  false  false  false
 false   true  false  false   true   true  false  false
  true  false  false  false   true  false  false  false
  true   true   true   true  false   true   true  false
 false  false   true  false   true  false   true   true
 false  false  false  false   true   true  false   true
 false  false  false  false  false   true   true  false

In [26]:
r = dijkstra_shortest_paths(g, dist, 1)

Graphs.DijkstraStates{Int64,Float64,DataStructures.MutableBinaryHeap{Graphs.DijkstraHEntry{Int64,Float64},DataStructures.LessThan},Int64}([1,1,5,1,2,3,5,7],[1,1,5,1,2,3,5,7],[0.0,9.0,22.0,13.0,12.0,26.0,33.0,38.0],[2,2,2,2,2,2,2,2],MutableBinaryHeap(),[0,1,4,3,2,5,6,7])

In [27]:
r.dists

8-element Array{Float64,1}:
  0.0
  9.0
 22.0
 13.0
 12.0
 26.0
 33.0
 38.0

^ These are the shortest distances from A to each letter.
- A => A is 0
- A => B is 9
- A => C is 22
- A => D is 13  
.  
.  
.  
- A => H is 38

Which are all the same shortest paths as the tropical math solution

### Problem 4:
This problem finds the shortest distance from Fitchburg State to the Entertainment Cinemas in Leominster
which mimics Google Maps or another mapping/GPS program.

**( a ) Create a graph that has Fitchburg State as the starting point (use 160 Pearl St.) and the Entertainment
Cinemas (45 Sack Blvd, Leominster, MA 01453) as the end. Include the following intersections in your
graph:**
- North Street and Main Street (Fitchburg)
- John Fitch Hwy and Lunenburg Street (Fitchburg)
- South Street and Laurel Street (Fitchburg)
- South Street and Wanoosnoc Rd (Fitchburg)
- Water Street and Bemis Rd (Fitchburg)
- Summer Street and John Fitch Hwy (Fitchburg)
- North Street and Main Street (Leominster)
- Main Street and Prospect Street (Leominster)
- N. Main Street and Rte 2 (Leominster)
- Merriam Ave and Rte 2 (Leominster)
- Haws Street and Commercial Road (Leominster)

(I also added two more intersections)
- Main St & Mill St, Leominster MA
- Mill St & Cinema Blvd, Leominster MA

In [74]:
A = render_map("cinemas_q1.json");
B = render_map("cinemas_q2.json");
C = render_map("cinemas_q3.json");
D = render_map("cinemas_q4.json");
the_map = 0.000621371*[A B; C D]

15×15 Array{Float64,2}:
 0.0       0.472863  0.695314  0.818346  …  4.81128   5.82722   5.85145 
 0.472863  0.0       0.840094  0.300122     4.29367   5.30899   5.33323 
 0.695314  0.840094  0.0       1.14084      4.33655   5.35249   5.37672 
 0.797219  0.300744  1.14084   0.0          4.2247    5.58861   5.61284 
 2.55881   2.06171   2.28665   1.76097      2.99252   3.82765   3.85188 
 2.35313   1.64601   1.6572    1.61743   …  2.69675   3.71207   3.73693 
 1.77526   1.11909   1.07932   1.42667      3.26282   4.27814   4.30237 
 3.39393   2.73714   2.69799   3.04472      2.89062   2.67997   2.70421 
 5.19901   4.54284   4.5037    5.54139      1.07373   0.863084  0.887318
 3.77918   3.2827    3.33614   3.19447      2.37985   2.16921   2.19344 
 3.34919   2.85209   3.07703   2.55197   …  2.37488   2.77069   3.04099 
 5.8558    5.3587    4.82867   5.05796      0.828288  0.358531  0.382765
 4.80071   4.30362   4.35705   4.21538      0.0       0.395813  0.665488
 6.21433   5.71723   5.1872

**( b ) Use the Tropical Matrix Arithmetic method to solve this problem. List the path travelled that minimizes the travel distance and the nodes (intersections) travelled through.**

In [76]:
the_map = trop_pow(the_map,3)

15×15 Array{Float64,2}:
 0.0       0.472863  0.695314  0.772986  …  4.76654   5.16235   5.43203 
 0.472863  0.0       0.840094  0.300122     4.29367   4.68949   4.95916 
 0.695314  0.840094  0.0       1.14022      4.33655   4.73236   5.00204 
 0.773607  0.300744  1.14084   0.0          4.2247    4.62051   4.89019 
 2.53457   2.06171   2.28602   1.76097      2.99252   3.38834   3.65801 
 2.11888   1.64601   1.6572    1.61743   …  2.69675   3.09256   3.36224 
 1.59195   1.11909   1.07932   1.41921      3.26282   3.65863   3.92831 
 3.21      2.73714   2.69799   3.03726      2.89062   2.67997   2.70421 
 5.01571   4.54284   4.50308   4.53352      1.07373   0.863084  0.887318
 3.75557   3.2827    3.33614   3.19447      2.37985   2.16921   2.19344 
 3.32496   2.85209   3.07641   2.55135   …  2.37488   2.77069   3.04037 
 5.18037   4.70751   4.76094   4.61927      0.754344  0.358531  0.382765
 4.77648   4.30362   4.35705   4.21538      0.0       0.395813  0.665488
 5.17229   4.69943   4.7528

In [77]:
the_map[:,1:7]

15×7 Array{Float64,2}:
 0.0       0.472863  0.695314  0.772986  2.53395   2.06979   1.59817 
 0.472863  0.0       0.840094  0.300122  2.06109   1.59692   1.1253  
 0.695314  0.840094  0.0       1.14022   2.27235   1.88773   1.07435 
 0.773607  0.300744  1.14084   0.0       1.76097   1.52795   1.42605 
 2.53457   2.06171   2.28602   1.76097   0.0       0.628827  1.31234 
 2.11888   1.64601   1.6572    1.61743   0.631934  0.0       0.683508
 1.59195   1.11909   1.07932   1.41921   1.198     0.813996  0.0     
 3.21      2.73714   2.69799   3.03726   2.68867   2.30467   1.61867 
 5.01571   4.54284   4.50308   4.53352   3.54803   2.91609   3.42375 
 3.75557   3.2827    3.33614   3.19447   2.27732   1.67894   2.36183 
 3.32496   2.85209   3.07641   2.55135   0.790384  1.41921   2.10272 
 5.18037   4.70751   4.76094   4.61927   3.29699   3.10375   3.78663 
 4.77648   4.30362   4.35705   4.21538   3.0298    2.69986   3.38337 
 5.17229   4.69943   4.75287   4.61119   3.42562   3.09567   3.7791

In [78]:
the_map[:,8:15]

15×8 Array{Float64,2}:
 3.20814  5.02503   6.82638  4.39496  5.42333   4.76654   5.16235   5.43203 
 2.73528  4.55216   6.35352  3.92209  4.95046   4.29367   4.68949   4.95916 
 2.6837   4.50059   6.39701  4.13336  4.99396   4.33655   4.73236   5.00204 
 3.03602  4.85291   6.28517  3.62197  4.88149   4.2247    4.62051   4.89019 
 2.68929  3.4803    4.87217  1.86101  3.46911   2.99252   3.38834   3.65801 
 2.06047  3.36535   4.75722  2.49294  3.35354   2.69675   3.09256   3.36224 
 1.8231   3.46663   5.32266  3.05901  3.91961   3.26282   3.65863   3.92831 
 0.0      1.81689   3.7245   4.54968  2.32144   2.89062   2.67997   2.70421 
 1.81689  0.0       1.90761  2.98942  0.504553  1.07373   0.863084  0.887318
 3.63875  1.82186   0.0      2.82165  1.81068   2.37985   2.16921   2.19344 
 3.47968  2.58552   4.49313  0.0      3.09008   2.37488   2.77069   3.04037 
 2.33014  0.513252  1.4248   2.50661  0.0       0.754344  0.358531  0.382765
 2.89062  1.07373   2.17915  2.23942  0.754344  0.0  

The shortest distance from *Fitchburg State* to *Entertainment Cinemas* is 5.43 miles

`A => B => D => F => M => N => O` based on the list in **`(C)`**

`FSU => Main St & North Street, Fitchburg => Laurel St & South St, Fitchburg => Water St & Bemis Rd, Fitchburg => Main St & Mill St, Leominster => Mill St & Cinema Blvd, Leominster => Entertainment Cinemas`

**( c ) Use Dijkstra’s method to solve the problem. List the path travelled that minimizes the travel distance and the nodes (intersections) travelled through.**

In [50]:
A = 1   # 160 Pearl Street, Fitchburg MA (Fitchburg State)
B = 2   # Main St & North Street, Fitchburg MA
C = 3   # Lunenburg St & John Fitch Hwy, Fitchburg MA
D = 4   # Laurel St & South St, Fitchburg MA
E = 5   # South St & Wanoosnoc Rd, Fitchburg MA
F = 6   # Water St & Bemis Rd, Fitchburg MA
G = 7   # John Fitch Hwy & Summer St, Fitchburg MA
H = 8   # Main St & North St, Leominster MA
I = 9   # Main Street & Prospect St, Leominster MA
J = 10  # N Main St & Rte 2, Leominster MA

K = 11  # Merriam Ave & Rte 2, Leominster MA
L = 12  # Commercial Rd & Haws St, Leominster MA
M = 13  # Main St & Mill St, Leominster MA
N = 14  # Mill St & Cinema Blvd, Leominster MA
O = 15  # 45 Sack Blvd, Leominster MA (Entertainment Cinemas)

cinema_map = Array{Tuple{Int, Int, Float64}}(0)

cinema_map = [
    (A, B, the_map[A,B]),  # Fitchburg State                      => Main @ North St, Fitchburg
    (B, C, the_map[B,C]),  # Main @ North St, Fitchburg           => Lunenburg Rd @ John Fitch Hwy, Fitchburg
    (B, D, the_map[B,D]),  # Main @ North St, Fitchburg           => Laurel St @ South St, Fitchburg
    (B, E, the_map[B,E]),  # Main @ North St, Fitchburg           => South St & Wanoosnoc Rd, Fitchburg
    (C, G, the_map[C,G]),  # Lunenburg Rd @ John Fitch, Fitchburg => John Fitch @ Summer St, Fitchburg
    (D, E, the_map[D,E]),  # Laurel St @ South St, Fitchburg      => South St & Wanoosnoc Rd, Fitchburg
    (E, D, the_map[E,D]),  # South St & Wanoosnoc Rd, Fitchburg   => Laurel St @ South St, Fitchburg
    (E, F, the_map[E,F]),  # South St & Wanoosnoc Rd, Fitchburg   => Water St @ Bemis Rd, Fitchburg
    (E, K, the_map[E,K]),  # South St & Wanoosnoc Rd, Fitchburg   => Merriam Ave @ Rte 2, Leominster
    (F, E, the_map[F,E]),  # Water St @ Bemis Rd, Fitchburg       => South St & Wanoosnoc Rd, Fitchburg
    (F, G, the_map[F,G]),  # Water St @ Bemis Rd, Fitchburg       => John Fitch @ Summer St, Fitchburg
    (F, J, the_map[F,J]),  # Water St @ Bemis Rd, Fitchburg       => N Main St @ Rte 2, Leominster
    (F, M, the_map[F,M]),  # Water St @ Bemis Rd, Fitchburg       => Main St @ Mill St, Leominster
    (G, F, the_map[G,F]),  # John Fitch @ Summer St, Fitchburg    => Water St @ Bemis Rd, Fitchburg
    (G, H, the_map[G,H]),  # John Fitch @ Summer St, Fitchburg    => Main St @ North St, Leominster
    (H, I, the_map[H,I]),  # Main St @ North St, Leominster       => Main St @ Prospect St, Leominster
    (I, L, the_map[I,L]),  # Main St @ Prospect St, Leominster    => Commercial Rd & Haws St, Leominster
    (J, L, the_map[J,L]),  # N Main St @ Rte 2, Leominster        => Commercial Rd & Haws St, Leominster
    (K, L, the_map[K,L]),  # Merriam Ave @ Rte 2, Leominster      => Commercial Rd & Haws St, Leominster
    (L, O, the_map[L,O]),  # Commercial Rd & Haws St, Leominster  => Entertainment Cinemas
    (M, N, the_map[M,N]),  # Main St @ Mill St, Leominster        => Mill St & Cinema Blvd, Leominster
    (N, O, the_map[N,O]),  # Mill St & Cinema Blvd, Leominster    => Entertainment Cinemas
]



# Ignoring this block, it didn't make sense for all nodes to be connected
#=
for i=1:13
    for j=1:13
        if the_map[i,j] != 0
            push!(cinema_map,(i,j,the_map[i,j]))
        end
    end
end
=#

cinema_map

22-element Array{Tuple{Int64,Int64,Float64},1}:
 (1,2,0.472863)  
 (2,3,0.840094)  
 (2,4,0.300122)  
 (2,5,2.06109)   
 (3,7,1.07435)   
 (4,5,1.76097)   
 (5,4,1.76097)   
 (5,6,0.628827)  
 (5,11,1.86101)  
 (6,5,0.631934)  
 (6,7,0.683508)  
 (6,10,4.75722)  
 (6,13,2.69675)  
 (7,6,0.813996)  
 (7,8,1.8231)    
 (8,9,1.81689)   
 (9,12,0.504553) 
 (10,12,1.81068) 
 (11,12,3.09008) 
 (12,15,0.382765)
 (13,14,0.395813)
 (14,15,0.269675)

In [51]:
dist_cinema_map = map(a->a[3],cinema_map)

22-element Array{Float64,1}:
 0.472863
 0.840094
 0.300122
 2.06109 
 1.07435 
 1.76097 
 1.76097 
 0.628827
 1.86101 
 0.631934
 0.683508
 4.75722 
 2.69675 
 0.813996
 1.8231  
 1.81689 
 0.504553
 1.81068 
 3.09008 
 0.382765
 0.395813
 0.269675

In [52]:
g2 = simple_inclist(15,is_directed=true)
for edge in cinema_map
    add_edge!(g2,edge[1],edge[2])
end
g2

Directed Graph (15 vertices, 22 edges)

In [53]:
adjacency_matrix(g2)

15×15 Array{Bool,2}:
 false   true  false  false  false  …  false  false  false  false  false
 false  false   true   true   true     false  false  false  false  false
 false  false  false  false  false     false  false  false  false  false
 false  false  false  false   true     false  false  false  false  false
 false  false  false   true  false      true  false  false  false  false
 false  false  false  false   true  …  false  false   true  false  false
 false  false  false  false  false     false  false  false  false  false
 false  false  false  false  false     false  false  false  false  false
 false  false  false  false  false     false   true  false  false  false
 false  false  false  false  false     false   true  false  false  false
 false  false  false  false  false  …  false   true  false  false  false
 false  false  false  false  false     false  false  false  false   true
 false  false  false  false  false     false  false  false   true  false
 false  false  false  false  f

In [54]:
r2 = dijkstra_shortest_paths(g2, dist_cinema_map, 1)

Graphs.DijkstraStates{Int64,Float64,DataStructures.MutableBinaryHeap{Graphs.DijkstraHEntry{Int64,Float64},DataStructures.LessThan},Int64}([1,1,2,2,4,5,3,7,8,6,5,9,6,13,14],[1,1,2,2,4,5,3,7,8,6,5,9,6,13,14],[0.0,0.472863,1.31296,0.772986,2.53395,3.16278,2.38731,4.21041,6.0273,7.91999,4.39496,6.53185,5.85953,6.25534,6.52502],[2,2,2,2,2,2,2,2,2,2,2,2,2,2,2],MutableBinaryHeap(),[0,1,2,3,4,6,5,7,11,9,8,12,10,13,14])

In [55]:
r2.dists

15-element Array{Float64,1}:
 0.0     
 0.472863
 1.31296 
 0.772986
 2.53395 
 3.16278 
 2.38731 
 4.21041 
 6.0273  
 7.91999 
 4.39496 
 6.53185 
 5.85953 
 6.25534 
 6.52502 